### DISCUSSION TITLE
Multi UE support features

### DISCUSSION THREAD

**Discussion by kalyan10** [srsran/srsRAN_Project#40] [General Help] [answered]
Hi,

Iam interested in multiUE testing. How many UE the srsranproject will support. If you give some more additional inputs on the mentioned area will helpful.

Thanks

kalyan

**Comment by ismagom** [comment, marked as answer]
Hi @kalyan10, 

We have tested up to 64 UEs in connected state using Amarisoft UE simulator.

**Reply by arrowtem** [reply]
> Hi @kalyan10,
> 
> We have tested up to 64 UEs in connected state using Amarisoft UE simulator.

Hello
Thank you much for srsRAN project and whole your team .
Is it technically possible to run multiple UE on 1 sdr ? ( I have b200 mini ) 
I saw amarisoft equipment , how did they managed to run that much UE’s having significantly less SDR’s?

**Reply by frankist** [reply]
With Amarisoft UE simulator, you can run multiple UEs using a single SDR and with a single binary run. You can set multiple UEs in the amarisoft ue config file.

**Reply by arrowtem** [reply]
how is it possible to run multiple ue’s on one sdr if PUCCH resources of several UE’s could be in same frequency place ? How they will send simultaneous ack/nack/sr/cqi etc ? 
the same with RACH 
Thanks for answer

**Reply by Zplusless** [reply]
Hi @ismagom , you mentioned that srsRAN works well with 64 UEs using the Amarisoft UE simulator. I'm wondering which setup is used (over the air or zmq). Can ZMQ-based setup support the same number of (or more?) UEs simultaneously?

**Reply by ismagom** [reply]
It works with both, zmq or air. Note with zmq you don't need to use a broker with amariUE since it simulates all UEs with a single PHY

**Reply by Zplusless** [reply]
@ismagom Thanks! If Amarisoft UE is used, the 20MHz channel bandwidth limitation of srsUE no longer exists. I'm wondering what is the maximum throughput per base station? Limited by the hardware?

**Comment by kalyan10** [comment]
Hi,
Thanks for your reply
if we want to test without Radio. with srsue how many we can do.

Thanks
kalyan

**Reply by mdalgitsis** [reply]
Hello @kalyan10 , if I am not wrong only 1 srsue can be registered when no real radio is used. For this the zeroMQ mode must be enabled. Then with GNU Radio multiple UEs can be registered as well. However, the real question is if we can use zeroMQ and srsue with the SRSRAN project.

Someone can answer that?

**Reply by mdalgitsis** [reply]
Ok, I think I got my answer here https://github.com/srsran/srsRAN_Project/issues/32

**Reply by jalbrethsen** [reply]
Has anyone actually got multiple ZMQ srsUE working with the gNB + open5gs? I use the GNU radio for multiple UE, but only one UE actually gets assigned an IP address, the other always gets stuck at 'RRC Connected'. Both have the same c-RNTI after 'Random Access Complete.' which makes me think the UE contention is not being handled correctly.

**Reply by ChengChen2020** [reply]
Hi @jalbrethsen, when I tried to connect multiple ZMQ srsUE, the others cannot go to the random access stage and stuck at "Attaching UE...". Have you solved the problem now? I have several questions about the configuration.
- Are you using the same USIM for all UEs?
- Are you creating different netns for different UEs?
- Have you tried creating multiple gNBs?
Thanks!

**Reply by jalbrethsen** [reply]
@ChengChen2020 I never got it working well and I haven't touched it in a few months, I did have different USIM and netns. If I remember correct, the ZMQ REQ/REPL sockets that srsRAN uses for virtual radio seem to only allow a single connection, to have multiple you will need to use a broker. I tried using GNU radio like in https://openaicellular.github.io/oaic/multi_ue_example.html which gave me the error I listed above. I also tried a custom broker using ZMQ dealer sockets and pub/sub sockets but I never got it to be very stable, as soon as I added a second UE the first one would break.

**Comment by kalyan10** [comment]
I tried only one UE to one gNB.. but many more gNB and UE to each

**Reply by nishaadh** [reply]
Do you mean you successfully tested multiple "one UE to one gnB" scenario? 
Are the gnBs using the same core network?
If so, what changes did you make enable that support?

**Comment by kalyan10** [comment]
Yes, I did!

**Reply by nishaadh** [reply]
Thank you for responding. If it is not too much trouble, would you mind pointing me to the right direction with this matter? Is there any documentation I can reference for attaching two gNB to one Open5GS core? Currently, I am having issue with both gNB trying to bind to the same amf address included on the open5gs config files and while the first attaches, the second one does not. How did you deal with this issue?

Any feedback is greatly appreciated!

**Reply by Aniket621** [reply]
Hi Nishaadh one question e u using both gNB on same machine with ZeroMQ.

**Reply by nishaadh** [reply]
I have not been able to set up more than one gNB as of now, but I was trying to set it up in the same machine. Do you have any information on if it is possible?

**Reply by natanzi** [reply]
I have the same issue, And I would like to run multiple gNB and Ue in the same machine, to run the handover scenario, can anyone point me to the right direction with this matter?

**Reply by natanzi** [reply]
Also multiple ue on multiple cell

**Comment by Zplusless** [comment]
Hi everyone. I tried 1gnb(release 23.10) + 2UEs(release 23.11) in zmq mode with gnuradio as a broker. Both UEs successfully get "RRC Connected" but only ue-2 set up PDU session and got an IP. The ue-2 works unstably with varying RTT (varying from 35ms to 4000+ms) and very low throughput (660kbps). I checked the log of ue-1 and find a "Network authentication failure". The configs and logs are listed as follows:

configs:
[ue_zmq-1.conf.txt](https://github.com/srsran/srsRAN_Project/files/13554914/ue_zmq-1.conf.txt)
[ue_zmq-2.conf.txt](https://github.com/srsran/srsRAN_Project/files/13554915/ue_zmq-2.conf.txt)
[gnb_zmq-multi.yaml.txt](https://github.com/srsran/srsRAN_Project/files/13554913/gnb_zmq-multi.yaml.txt)

gnu-radio:
[gnu-broker-2ue.zip](https://github.com/srsran/srsRAN_Project/files/13554918/gnu-broker-2ue.zip)

logs:
[gnb.log.txt](https://github.com/srsran/srsRAN_Project/files/13554922/gnb.log.txt)
[ue1.log.txt](https://github.com/srsran/srsRAN_Project/files/13554923/ue1.log.txt)
[ue2-iperf.txt](https://github.com/srsran/srsRAN_Project/files/13554924/ue2-iperf.txt)
[ue2-ping.txt](https://github.com/srsran/srsRAN_Project/files/13554925/ue2-ping.txt)
[ue2.log.txt](https://github.com/srsran/srsRAN_Project/files/13554928/ue2.log.txt)

Screenshot:
![剪贴板图片](https://github.com/srsran/srsRAN_Project/assets/10062702/0702395c-24df-44c3-bc50-1c211db3193b)
![image](https://github.com/srsran/srsRAN_Project/assets/10062702/61b7eebc-23f8-4b27-98ed-c6dea3588dbb)

**Reply by pgawlowicz** [reply]
@Zplusless Please check the new section of our gnb+srsUE tutorial that describes how to emulate multiple-UEs:
https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#multi-ue-emulation

Please remember to clone the newest version of the srsRAN-4G repository.
